torch/_higher_order_ops/scan.py

```diff
@@ -0,0 +1,402 @@
+        https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype
+
+    Args:
+        combine_fn (Callable): A binary callable with type ``(Tensor, Tensor) -> (Tensor, Tensor)``,
+            or if input is a pytree ``(pytree, pytree) -> (pytree, pytree)``.
+            The first input to ``combine_fn`` is the previous or initial scan carry
+            and the second input element to ``combine_fn`` is a slice of the input along dim.
+            The first output element of ``combine_fn`` is the next scan carry
+            and the second output  of ``combine_fn`` represents a slice of the output.
+            This function must be pure, i.e., no lifted arguments are supported at the moment
+            and may not have any side effects.
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors.
+            The ``init`` is expected to have the same pytree structure as the first output element
+            of ``combine_fn``.
+        input (torch.Tensor): The input tensor, or nested pytree of tensors.
+        dim (int): the dimension to scan over
+        reverse (bool): A boolean stating if the scan should be reversed with respect to ``dim``, default ``False``.
+
+
+    Example::
+
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y, x + y
+
+        i0 = 0.
+        x = [1., 2., 3., 4.]
+        cumsum = scan(add, init=i0, input=x, dim)
+        # This produces the output
+        carry, cumsum = 10., torch.stack([1., 3., 7., 10.], dim)
+
+
+    """
+    if not callable(combine_fn):
+        raise RuntimeError("Combine_fn must be a callable, but got {combine_fn}")
+    if not isinstance(dim, int):
+        raise RuntimeError("Dim must be an int, but got " + str(type(dim)))
+
+    # TODO: Support closures/nn_modules in order to be able represent RNNs with scan
+    # TODO: Support _inductor lowering
+    # TODO: Support Autograd
+    # TODO: Unify handling of pytrees for control flow ops, such as cond, while_loop, etc.
+
+    # Dynamo is expecting a callable with "__code__" attribute.
+    # We cannot directly pass cond_op to it. So we wrap it in a dummy function.
+    def _scan_op_wrapper(*args, **kwargs):
+        return scan(*args, **kwargs)
+
+    if not torch._dynamo.is_compiling():
+        with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
+            return torch.compile(_scan_op_wrapper, backend="eager", fullgraph=True)(
+                combine_fn, init=init, input=input, dim=dim, reverse=reverse
+            )
+
+    leaves_init, spec_init = pytree.tree_flatten(init)
+    leaves_input, spec_input = pytree.tree_flatten(input)
+
+    def check_arg(arg):
+        if any(not isinstance(x, torch.Tensor) for x in arg):
+            raise RuntimeError("All leaves must be a Tensor")
+        if any(x.shape[dim] == 0 for x in arg):
+            raise RuntimeError("All leaves must have a scan dimension > 0")
+
+    check_arg(leaves_init)
```

Comment:
Leaves of init don't necessarily have the same shape[dim], right? only input is required to have a > 0 dim. 

Do we have a test case for this?

Reply:
Correct, init does not have to have the same shape as input and also does not have to have shape[dim] > 0. I have remove this check for the init and also checked that we have a testcase for that. We had a [testcase](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L2435-L2533) for that already, but I added an additional one.